docs: Design document for the Remote Config (#839)
platform-platform/monorepo: metrics/web/docs/features/remote_config/01_remote_config_design.md

```

## Metrics application

### Domain layer

In the application domain layer, we should add an ability to read values from the `Remote Config`. For this purpose, we should:

1. Create the `RemoteConfig` entity.
2. Add the `getConfig` method to the `RemoteConfigRepository`.
3. Implement the `GetRemoteConfigUseCase` class.

So, the domain layer should look like this:

![Domain layer diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/remote_config_design/metrics/web/docs/features/remote_config/diagrams/remote_config_domain_class.puml)

### Data layer

The `FirebaseRemoteConfigRepository` of the data layer should implement the methods from the `RemoteConfigRepository` interface. To do that, we should create a `FirebaseRemoteConfigAdapter` class.

The following class diagram represents the classes of the data layer required for this feature:

![Data layer diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/remote_config_design/metrics/web/docs/features/remote_config/diagrams/remote_config_data_class.puml)

### Presentation layer

> Once we've created a domain and data layers, it's time to create a presentation layer. This layer contains the `RemoteConfigNotifier` - the class that manages `Remote Config` values.

Reviewer: ```suggestion
Once we've added both the domain and data layers, it's time to add the feature to the presentation. The state of the remote config is maintained by the `RemoteConfigNotifier` integrated to the `InjectionContainer` so it is available within the application. 
```

> 1. Create a `RemoteConfigNotifier` class.
> 2. Implement the fields that stand for `Remote Config` values.
> 3. Inject the notifier into the `InjectionContainer`.

Reviewer: ```suggestion
```

> The following class diagram represents the classes of the presentation layer required for this feature:

Reviewer: ```suggestion
The following class diagram demonstrates the structure of the presentation layer:
```

![Presentation layer diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/remote_config_design/metrics/web/docs/features/remote_config/diagrams/remote_config_presentation_class.puml)

> So, when the user enters the application, the following sequence diagram describes how the application applies `Remote Config` values:

Reviewer: ```suggestion
The following sequence diagram describes how the application applies `Remote Config` values when a user enters the application:
```

![Sequence diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/remote_config_design/metrics/web/docs/features/remote_config/diagrams/remote_config_sequence.puml)

> Let's consider the mechanism of applying the `Remote Config` values in the application.
> 
> When the user enters the application, he sees the `LoadingPage` while waiting until the `_initializeRemoteConfig` method of the `RemoteConfigNotifier` finishes. After that, the `isLoading` status of the `RemoteConfigNotifier` sets to `false`, and the user can see the UI with `Remote Config` values applied.

Reviewer: ```suggestion
Let's consider the mechanism of applying the `Remote Config` values in the application. When a user enters the application, he or she stays on the `LoadingPage` until the `_initializeRemoteConfig` method of the `RemoteConfigNotifier` finishes. Once the initializing completes, the `isLoading` status of the `RemoteConfigNotifier` is set to `false`. The user then can proceed to the application with the configurations applied. 
```
We should keep in mind here that the described behavior requires changes to the `LoadingPage` implementation. Maybe, it would be even a separate ticket as this seems to be not a small task.

### Data layer

The data layer provides the `FirebaseRemoteConfigurationRepository` implementation that uses the [Firebase package](https://pub.dev/packages/firebase) available for Flutter for Web and the `RemoteConfigurationData` that represents the `RemoteConfiguration` entity.

The following class diagram states the structure of the data layer:

![Data layer diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/remote_config_design/metrics/web/docs/features/remote_config/diagrams/remote_config_data_layer_class_diagram.puml)

### Domain layer

The domain layer should provide an interface for the `FirebaseRemoteConfigurationRepository` we need to interact with the `Remote Config API`. Also, the layer provides all the use cases required to interact with the repository, and entities required for the `Remote Config` feature. Thus, the following list of classes should be implemented to fit the feature requirements:

> - Implement the `RemoteConfigurationRepository` interface with a `applyDefaults()`, `fetch()`, `activate()`, and `getConfiguration()` methods.

Reviewer: ```suggestion
- Implement the `RemoteConfigurationRepository` interface with appropriate methods.
```

- Add the `RemoteConfiguration` entity with fields that come from a remote API.

> - Add the `FetchRemoteConfigurationUseCase` to perform fetching configurations.

Reviewer: Do we need the `SetDefaultRemoteConfigurationUseCase` as it seems to be a separate action for the application?
